src/supervisor.tex

```diff
@@ -2258,3 +2273,97 @@ \chapter{``Svnapot'' Standard Extension for NAPOT Translation Contiguity, Versio
+rules, and accesses to such pages are considered main memory rather than I/O
+for the purposes of FENCE, {\em.aq}, and {\em.rl}.
+
+If the underlying physical memory attribute for a page is main memory, then
+accesses to that page via a page table entry with PBMT=2 obey strong channel 0
+ordering with respect to other accesses to physical main memory and to other
+accesses to pages with PBMT=2.  Furthermore, accesses to such pages are
+considered I/O rather than main memory for the purposes of FENCE, {\em.aq}, and
+{\em.rl}.
+
+With Svpbmt enabled, it is possible for multiple virtual aliases of the same
+physical page to exist simultaneously with different memory attributes.  It is
+also possible for a U-mode or S-mode mapping through a PTE with Svpbmt enabled
+to observe different memory attributes for a given region of physical memory
+than a concurrent access to the same page performed by M-mode or when {\tt
+satp}.MODE=Bare.  If accesses are performed simultaneously to the same region
+of memory using different attributes, there may be a loss of coherence and/or
+of normal RVWMO, RVTSO, or I/O ordering semantics.  In such cases,
+platform-specific mechanisms must be used to restore corrent coherence and
+memory ordering.
+
+\begin{commentary}
+For example, a cacheable access may be issued at the same time as a
+non-cacheable access to the same physical memory address.  In this case,
+if the former is performed first in the global memory order, then it will
+be evicted from the cache by the latter.  If on the other hand the cacheable
+access appears after the non-cacheable access, then the former may remain
+cached as it normally would.
+
+Likewise, accesses performed under memory indicating the non-idempotent
+attribute must not be merged with idempotent accesses to the same region
+in flight at the same time, as the non-idempotency of the former must
+be respected.  This is not expected to be a common situation.
+
+Note that Svpbmt cannot be used to completely prevent speculative reads from
+being performed to a region of memory for which the PMAs indicate idempotence,
+as speculation can still be performed via M-mode or via Bare mappings, which do
+not use the PBMTs.  Platform-specific mechanisms must be used to avoid this
+form of conflict.
+\end{commentary}
+
+\begin{commentary}
+A device driver written to rely on I/O strong ordering rules will not
+operate correctly if the address range is mapped as main memory by the
+page-based memory types.  As such, this configuration is discouraged.
+
+In spite of this caveat, it will often still be useful to map physical I/O
+regions using PBMT=1 so that write combining and speculative accesses can be
+performed.  Such optimizations will likely improve performance when applied
+with adequate care.
+\end{commentary}
+
+When two-level paging is enabled within the H extension, the page-based memory
+types are applied in two stages.  First, the G-stage PTE PBMT bits (if enabled)
```

Comment:
What does "if enabled" mean here? Is there some global enable for Svpbmt? Or does this mean "if the PTE's PBMT bits are nonzero"?

Reply:
I meant "if hgatp.MODE != 0" and "vsatp.MODE != 0", because otherwise there are no PTE bits to consider at all, but I will reword to clarify.